packages/babel-plugin-minify-dead-code-elimination/src/index.js

```diff
@@ -238,10 +247,25 @@ module.exports = ({ types: t, traverse }) => {
             }
 
             if (binding.references === 1 && binding.kind !== "param" && binding.kind !== "module" && binding.constant) {
+
               let replacement = binding.path.node;
               let replacementPath = binding.path;
+              let isAfterReturn = false;
+
               if (t.isVariableDeclarator(replacement)) {
-                replacement = replacement.init;
+
+                forEachPrevSibling(replacementPath, (path) => {
+                  if (t.isReturnStatement(path)) {
+                    isAfterReturn = true;
+                  }
+                });
+
+                // simulate hoisting by replacing value
+                // with undefined if declaration is after return
+                replacement = isAfterReturn
+                  ? t.identifier("undefined")
+                  : replacement.init;
+
```

Comment:
a few lines below this, `replacementPath = replacementPath.get("init")` - we might have to fix this here too. Else replacement and replacementPath might point to different things.